File: tests/performance_tests/k6_scripts/api_load_test.js
import http from 'k6/http';
import { check, group, sleep } from 'k6';
import { Rate } from 'k6/metrics';
import { randomString, randomIntBetween } from 'https://jslib.k6.io/k6-utils/1.2.0/index.js';

// Custom metrics
const errorRate = new Rate('errors');

// Default options
export const options = {
  vus: 50,              // Number of virtual users
  duration: '1m',       // Test duration
  thresholds: {
    http_req_duration: ['p(95)<1000'], // 95% of requests should be below 1s
    errors: ['rate<0.01'],             // Error rate should be less than 1%
  },
  stages: [
    { duration: '10s', target: 50 },   // Ramp-up to 50 users over 10 seconds
    { duration: '40s', target: 50 },   // Stay at 50 users for 40 seconds
    { duration: '10s', target: 0 },    // Ramp-down to 0 users over 10 seconds
  ],
};

// Get base URL from environment variable or use default
const BASE_URL = __ENV.BASE_URL || 'https://api-dev.example.com';

// Shared parameters
const THINK_TIME = parseFloat(__ENV.THINK_TIME || '1');

// Utility function to generate random user data
function generateUserData() {
  return {
    name: `Test User ${randomString(5)}`,
    email: `test.${randomString(8)}@example.com`,
    phone: `+1${randomIntBetween(2, 9)}${randomIntBetween(0, 9)}${randomString(8, '0123456789')}`,
  };
}

// Utility function to generate random product data
function generateProductData() {
  return {
    name: `Test Product ${randomString(5)}`,
    description: 'A test product created via performance tests',
    price: parseFloat((randomIntBetween(10, 200) + 0.99).toFixed(2)),
    category: 'test',
    stock: randomIntBetween(50, 500),
    sku: `TEST-${randomString(6, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789')}`,
  };
}

// Utility function to generate random order data
function generateOrderData() {
  return {
    customer_id: randomIntBetween(100, 999),
    items: [
      {
        product_id: randomIntBetween(1, 10),
        quantity: randomIntBetween(1, 5),
        price: parseFloat((randomIntBetween(10, 100) + 0.99).toFixed(2)),
      },
      {
        product_id: randomIntBetween(11, 20),
        quantity: randomIntBetween(1, 3),
        price: parseFloat((randomIntBetween(20, 200) + 0.99).toFixed(2)),
      },
    ],
    shipping_address: {
      street: `${randomIntBetween(100, 999)} Main St`,
      city: 'Test City',
      state: 'TS',
      zip: randomString(5, '0123456789'),
      country: 'Test Country',
    },
  };
}

// Main test function
export default function() {
  let userId, productId, orderId;

  // Common headers
  const headers = {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
  };

  // Users API tests
  group('Users API', () => {
    // GET All Users
    let getAllUsersResponse = http.get(`${BASE_URL}/users`, { headers });
    
    check(getAllUsersResponse, {
      'GET All Users status is 200': (r) => r.status === 200,
      'GET All Users has valid JSON response': (r) => r.json().length > 0,
    }) || errorRate.add(1);

    // GET User by ID
    let getUserResponse = http.get(`${BASE_URL}/users/1`, { headers });
    
    check(getUserResponse, {
      'GET User status is 200': (r) => r.status === 200,
      'GET User has correct ID': (r) => r.json().id === 1,
    }) || errorRate.add(1);

    // POST Create User
    let userData = generateUserData();
    let createUserResponse = http.post(
      `${BASE_URL}/users`,
      JSON.stringify(userData),
      { headers }
    );
    
    check(createUserResponse, {
      'POST Create User status is 201': (r) => r.status === 201,
      'POST Create User has valid ID': (r) => r.json().id !== undefined,
    }) || errorRate.add(1);

    if (createUserResponse.status === 201) {
      userId = createUserResponse.json().id;
    }

    sleep(THINK_TIME);
  });

  // Products API tests
  group('Products API', () => {
    // GET All Products
    let getAllProductsResponse = http.get(`${BASE_URL}/products`, { headers });
    
    check(getAllProductsResponse, {
      'GET All Products status is 200': (r) => r.status === 200,
      'GET All Products has valid JSON response': (r) => r.json().length > 0,
    }) || errorRate.add(1);

    if (getAllProductsResponse.status === 200 && getAllProductsResponse.json().length > 0) {
      productId = getAllProductsResponse.json()[0].id;
    }

    // GET Product by ID
    if (productId) {
      let getProductResponse = http.get(`${BASE_URL}/products/${productId}`, { headers });
      
      check(getProductResponse, {
        'GET Product status is 200': (r) => r.status === 200,
        'GET Product has correct ID': (r) => r.json().id === productId,
      }) || errorRate.add(1);
    }

    // GET Products by Category
    let getProductsByCategoryResponse = http.get(
      `${BASE_URL}/products?category=electronics`,
      { headers }
    );
    
    check(getProductsByCategoryResponse, {
      'GET Products by Category status is 200': (r) => r.status === 200,
      'GET Products by Category has valid category': (r) => 
        r.json().length > 0 && r.json()[0].category === 'electronics',
    }) || errorRate.add(1);

    // POST Create Product
    let productData = generateProductData();
    let createProductResponse = http.post(
      `${BASE_URL}/products`,
      JSON.stringify(productData),
      { headers }
    );
    
    check(createProductResponse, {
      'POST Create Product status is 201': (r) => r.status === 201,
      'POST Create Product has valid ID': (r) => r.json().id !== undefined,
    }) || errorRate.add(1);

    sleep(THINK_TIME);
  });

  // Orders API tests
  group('Orders API', () => {
    // GET All Orders
    let getAllOrdersResponse = http.get(`${BASE_URL}/orders`, { headers });
    
    check(getAllOrdersResponse, {
      'GET All Orders status is 200': (r) => r.status === 200,
      'GET All Orders has valid JSON response': (r) => r.json().length > 0,
    }) || errorRate.add(1);

    if (getAllOrdersResponse.status === 200 && getAllOrdersResponse.json().length > 0) {
      orderId = getAllOrdersResponse.json()[0].id;
    }

    // GET Order by ID
    if (orderId) {
      let getOrderResponse = http.get(`${BASE_URL}/orders/${orderId}`, { headers });
      
      check(getOrderResponse, {
        'GET Order status is 200': (r) => r.status === 200,
        'GET Order has correct ID': (r) => r.json().id === orderId,
      }) || errorRate.add(1);
    }

    // POST Create Order
    let orderData = generateOrderData();
    let createOrderResponse = http.post(
      `${BASE_URL}/orders`,
      JSON.stringify(orderData),
      { headers }
    );
    
    check(createOrderResponse, {
      'POST Create Order status is 201': (r) => r.status === 201,
      'POST Create Order has valid ID': (r) => r.json().id !== undefined,
    }) || errorRate.add(1);

    sleep(THINK_TIME);
  });
}

File: tests/performance_tests/locust_scripts/api_load_test.py
import time
import json
import random
import string
import os
from locust import HttpUser, task, between, tag

class ApiUser(HttpUser):
    """
    Simulates a user interacting with the API
    """
    # Wait between 1 and 5 seconds between tasks
    wait_time = between(1, 5)
    
    # Store IDs for later use
    user_id = None
    product_id = None
    order_id = None
    
    def on_start(self):
        """
        Initialize the user session
        """
        # Set default headers
        self.client.headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        }
    
    def _random_string(self, length=8, chars=string.ascii_lowercase + string.digits):
        """Generate a random string"""
        return ''.join(random.choice(chars) for _ in range(length))
    
    def _random_phone(self):
        """Generate a random phone number"""
        return f"+1{random.randint(2, 9)}{random.randint(0, 9)}{''.join(str(random.randint(0, 9)) for _ in range(8))}"
    
    def _generate_user_data(self):
        """Generate random user data"""
        return {
            "name": f"Test User {self._random_string(5)}",
            "email": f"test.{self._random_string(8)}@example.com",
            "phone": self._random_phone()
        }
    
    def _generate_product_data(self):
        """Generate random product data"""
        return {
            "name": f"Test Product {self._random_string(5)}",
            "description": "A test product created via performance tests",
            "price": round(random.uniform(10, 200), 2),
            "category": "test",
            "stock": random.randint(50, 500),
            "sku": f"TEST-{self._random_string(6, string.ascii_uppercase + string.digits)}"
        }
    
    def _generate_order_data(self):
        """Generate random order data"""
        return {
            "customer_id": random.randint(100, 999),
            "items": [
                {
                    "product_id": random.randint(1, 10),
                    "quantity": random.randint(1, 5),
                    "price": round(random.uniform(10, 100), 2)
                },
                {
                    "product_id": random.randint(11, 20),
                    "quantity": random.randint(1, 3),
                    "price": round(random.uniform(20, 200), 2)
                }
            ],
            "shipping_address": {
                "street": f"{random.randint(100, 999)} Main St",
                "city": "Test City",
                "state": "TS",
                "zip": ''.join(random.choice(string.digits) for _ in range(5)),
                "country": "Test Country"
            }
        }
    
    @tag('users')
    @task(3)
    def get_all_users(self):
        """Get all users"""
        with self.client.get("/users", name="GET All Users") as response:
            if response.status_code == 200:
                users = response.json()
                if users and len(users) > 0:
                    # Store a user ID for later use
                    self.user_id = users[0]["id"]
    
    @tag('users')
    @task(4)
    def get_user_by_id(self):
        """Get a specific user by ID"""
        # Use the stored user ID or default to 1
        user_id = self.user_id or 1
        self.client.get(f"/users/{user_id}", name="GET User by ID")
    
    @tag('users')
    @task(2)
    def create_user(self):
        """Create a new user"""
        user_data = self._generate_user_data()
        with self.client.post("/users", json=user_data, name="POST Create User") as response:
            if response.status_code == 201:
                created_user = response.json()
                self.user_id = created_user["id"]
    
    @tag('products')
    @task(3)
    def get_all_products(self):
        """Get all products"""
        with self.client.get("/products", name="GET All Products") as response:
            if response.status_code == 200:
                products = response.json()
                if products and len(products) > 0:
                    # Store a product ID for later use
                    self.product_id = products[0]["id"]
    
    @tag('products')
    @task(4)
    def get_product_by_id(self):
        """Get a specific product by ID"""
        # Use the stored product ID or default to P001
        product_id = self.product_id or "P001"
        self.client.get(f"/products/{product_id}", name="GET Product by ID")
    
    @tag('products')
    @task(2)
    def get_products_by_category(self):
        """Get products by category"""
        self.client.get("/products?category=electronics", name="GET Products by Category")
    
    @tag('products')
    @task(2)
    def create_product(self):
        """Create a new product"""
        product_data = self._generate_product_data()
        with self.client.post("/products", json=product_data, name="POST Create Product") as response:
            if response.status_code == 201:
                created_product = response.json()
                self.product_id = created_product["id"]
    
    @tag('orders')
    @task(3)
    def get_all_orders(self):
        """Get all orders"""
        with self.client.get("/orders", name="GET All Orders") as response:
            if response.status_code == 200:
                orders = response.json()
                if orders and len(orders) > 0:
                    # Store an order ID for later use
                    self.order_id = orders[0]["id"]
    
    @tag('orders')
    @task(4)
    def get_order_by_id(self):
        """Get a specific order by ID"""
        # Use the stored order ID or default to ORD-001
        order_id = self.order_id or "ORD-001"
        self.client.get(f"/orders/{order_id}", name="GET Order by ID")
    
    @tag('orders')
    @task(2)
    def create_order(self):
        """Create a new order"""
        order_data = self._generate_order_data()
        with self.client.post("/orders", json=order_data, name="POST Create Order") as response:
            if response.status_code == 201:
                created_order = response.json()
                self.order_id = created_order["id"]
    
    @tag('orders')
    @task(1)
    def update_order_status(self):
        """Update an order status"""
        # Use the stored order ID or default to ORD-001
        order_id = self.order_id or "ORD-001"
        update_data = {
            "status": "shipped",
            "tracking_number": f"TRACK{self._random_string(8, string.digits)}"
        }
        self.client.patch(f"/orders/{order_id}", json=update_data, name="PATCH Update Order")

if __name__ == "__main__":
    # This is for running standalone, Locust will use its own command-line parsing when run from CLI
    from locust.env import Environment
    from locust.stats import stats_printer, stats_history
    from locust.log import setup_logging
    import gevent
    
    # Set up logging
    setup_logging("INFO", None)
    
    # Create an environment
    env = Environment(user_classes=[ApiUser])
    env.create_local_runner()
    
    # Start the test
    env.runner.start(50, spawn_rate=10)
    
    # Print stats in the console
    gevent.spawn(stats_printer(env.stats))
    
    # Save stats to history
    gevent.spawn(stats_history, env.runner)
    
    # Run for 60 seconds
    time.sleep(60)
    
    # Stop the test
    env.runner.quit()
